File: sdk/php/tests/ValidateTest.php
<?php

declare(strict_types=1);

namespace Cordum\Cap\Tests;

use Cordum\Agent\V1\BusPacket;
use Cordum\Agent\V1\JobRequest;
use Cordum\Agent\V1\JobResult;
use Cordum\Agent\V1\JobStatus;
use Cordum\Agent\V1\Budget;
use Cordum\Cap\Validate;
use Cordum\Cap\ValidationException;
use Google\Protobuf\Timestamp;
use PHPUnit\Framework\TestCase;

class ValidateTest extends TestCase
{
    #[\PHPUnit\Framework\Attributes\Test]
    public function validJobRequestPasses(): void
    {
        $req = new JobRequest();
        $req->setJobId('j-1');
        $req->setTopic('test');
        Validate::validateJobRequest($req);
        $this->assertTrue(true); // No exception
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function emptyJobIdRejected(): void
    {
        $req = new JobRequest();
        $req->setTopic('test');
        $this->expectException(ValidationException::class);
        Validate::validateJobRequest($req);
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function emptyTopicRejected(): void
    {
        $req = new JobRequest();
        $req->setJobId('j-1');
        $this->expectException(ValidationException::class);
        Validate::validateJobRequest($req);
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function validJobResultPasses(): void
    {
        $res = new JobResult();
        $res->setJobId('j-1');
        $res->setWorkerId('w-1');
        $res->setStatus(JobStatus::JOB_STATUS_SUCCEEDED);
        Validate::validateJobResult($res);
        $this->assertTrue(true);
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function unspecifiedStatusRejected(): void
    {
        $res = new JobResult();
        $res->setJobId('j-1');
        $res->setWorkerId('w-1');
        $this->expectException(ValidationException::class);
        Validate::validateJobResult($res);
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function emptyWorkerIdRejected(): void
    {
        $res = new JobResult();
        $res->setJobId('j-1');
        $res->setStatus(JobStatus::JOB_STATUS_SUCCEEDED);
        $this->expectException(ValidationException::class);
        Validate::validateJobResult($res);
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function validBusPacketPasses(): void
    {
        $req = new JobRequest();
        $req->setJobId('j');
        $req->setTopic('t');

        $ts = new Timestamp();
        $ts->setSeconds(1);

        $pkt = new BusPacket();
        $pkt->setTraceId('t');
        $pkt->setSenderId('s');
        $pkt->setProtocolVersion(1);
        $pkt->setCreatedAt($ts);
        $pkt->setJobRequest($req);

        Validate::validateBusPacket($pkt);
        $this->assertTrue(true);
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function emptySenderIdRejected(): void
    {
        $pkt = new BusPacket();
        $pkt->setTraceId('t');
        $pkt->setProtocolVersion(1);
        $this->expectException(ValidationException::class);
        Validate::validateBusPacket($pkt);
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function emptyTraceIdRejected(): void
    {
        $pkt = new BusPacket();
        $pkt->setSenderId('s');
        $pkt->setProtocolVersion(1);
        $this->expectException(ValidationException::class);
        Validate::validateBusPacket($pkt);
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function zeroProtocolVersionRejected(): void
    {
        $pkt = new BusPacket();
        $pkt->setTraceId('t');
        $pkt->setSenderId('s');
        $ts = new Timestamp();
        $ts->setSeconds(1);
        $pkt->setCreatedAt($ts);
        $this->expectException(ValidationException::class);
        Validate::validateBusPacket($pkt);
    }
}

File: sdk/java/src/test/java/io/cordum/cap/ValidateTest.java
package io.cordum.cap;

import com.google.protobuf.Timestamp;
import io.cordum.cap.agent.v1.BusPacket;
import io.cordum.cap.agent.v1.Budget;
import io.cordum.cap.agent.v1.JobPriority;
import io.cordum.cap.agent.v1.JobRequest;
import io.cordum.cap.agent.v1.JobResult;
import io.cordum.cap.agent.v1.JobStatus;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class ValidateTest {

    @Test
    void validJobRequest() {
        JobRequest req = JobRequest.newBuilder()
                .setJobId("job-1")
                .setTopic("test.topic")
                .build();
        assertDoesNotThrow(() -> Validate.validateJobRequest(req));
    }

    @Test
    void criticalJobPriorityIsAccepted() {
        JobRequest req = JobRequest.newBuilder()
                .setJobId("job-critical")
                .setTopic("test.topic")
                .setPriority(JobPriority.JOB_PRIORITY_CRITICAL)
                .build();
        assertDoesNotThrow(() -> Validate.validateJobRequest(req));
    }

    @Test
    void unknownJobPriorityIsRejected() {
        JobRequest req = JobRequest.newBuilder()
                .setJobId("job-unknown")
                .setTopic("test.topic")
                .setPriorityValue(JobPriority.JOB_PRIORITY_CRITICAL_VALUE + 1)
                .build();
        assertThrows(Errors.ValidationException.class,
                () -> Validate.validateJobRequest(req));
    }

    @Test
    void jobRequestNullRejects() {
        assertThrows(Errors.ValidationException.class,
                () -> Validate.validateJobRequest(null));
    }

    @Test
    void jobRequestEmptyJobIdRejects() {
        JobRequest req = JobRequest.newBuilder().setTopic("t").build();
        Errors.ValidationException ex = assertThrows(Errors.ValidationException.class,
                () -> Validate.validateJobRequest(req));
        assertEquals("job_id", ex.field());
    }

    @Test
    void jobRequestEmptyTopicRejects() {
        JobRequest req = JobRequest.newBuilder().setJobId("j").build();
        Errors.ValidationException ex = assertThrows(Errors.ValidationException.class,
                () -> Validate.validateJobRequest(req));
        assertEquals("topic", ex.field());
    }

    @Test
    void jobRequestNegativeBudgetRejects() {
        JobRequest req = JobRequest.newBuilder()
                .setJobId("j").setTopic("t")
                .setBudget(Budget.newBuilder().setMaxInputTokens(-1).build())
                .build();
        assertThrows(Errors.ValidationException.class,
                () -> Validate.validateJobRequest(req));
    }

    @Test
    void validJobResult() {
        JobResult res = JobResult.newBuilder()
                .setJobId("j-1")
                .setWorkerId("w-1")
                .setStatus(JobStatus.JOB_STATUS_SUCCEEDED)
                .build();
        assertDoesNotThrow(() -> Validate.validateJobResult(res));
    }

    @Test
    void jobResultUnspecifiedStatusRejects() {
        JobResult res = JobResult.newBuilder()
                .setJobId("j").setWorkerId("w").build();
        assertThrows(Errors.ValidationException.class,
                () -> Validate.validateJobResult(res));
    }

    @Test
    void jobResultEmptyWorkerRejects() {
        JobResult res = JobResult.newBuilder()
                .setJobId("j")
                .setStatus(JobStatus.JOB_STATUS_SUCCEEDED)
                .build();
        Errors.ValidationException ex = assertThrows(Errors.ValidationException.class,
                () -> Validate.validateJobResult(res));
        assertEquals("worker_id", ex.field());
    }

    @Test
    void validBusPacket() {
        BusPacket pkt = BusPacket.newBuilder()
                .setTraceId("t").setSenderId("s").setProtocolVersion(1)
                .setCreatedAt(Timestamp.newBuilder().setSeconds(1).build())
                .setJobRequest(JobRequest.newBuilder()
                        .setJobId("j").setTopic("t").build())
                .build();
        assertDoesNotThrow(() -> Validate.validateBusPacket(pkt));
    }

    @Test
    void busPacketEmptySenderRejects() {
        BusPacket pkt = BusPacket.newBuilder()
                .setTraceId("t").setProtocolVersion(1)
                .setCreatedAt(Timestamp.newBuilder().setSeconds(1).build())
                .setJobRequest(JobRequest.newBuilder()
                        .setJobId("j").setTopic("t").build())
                .build();
        Errors.ValidationException ex = assertThrows(Errors.ValidationException.class,
                () -> Validate.validateBusPacket(pkt));
        assertEquals("sender_id", ex.field());
    }

    @Test
    void busPacketNoPayloadRejects() {
        BusPacket pkt = BusPacket.newBuilder()
                .setTraceId("t").setSenderId("s").setProtocolVersion(1)
                .setCreatedAt(Timestamp.newBuilder().setSeconds(1).build())
                .build();
        Errors.ValidationException ex = assertThrows(Errors.ValidationException.class,
                () -> Validate.validateBusPacket(pkt));
        assertEquals("payload", ex.field());
    }
}
